python/ql/lib/semmle/python/ApiGraphs.qll

```diff
@@ -374,13 +503,55 @@ module API {
      *
      * `moduleImport("foo").getMember("bar")`
      */
-    private TApiNode potential_import_star_base(Scope s) {
+    private TApiNode potential_import_star_base(py::Scope s) {
       exists(DataFlow::Node n |
         n.asCfgNode() = ImportStar::potentialImportStarBase(s) and
         use(result, n)
       )
     }
 
+    /**
+     * Holds if `rhs` is the right-hand side of a definition of a node that should have an
+     * incoming edge from `base` labeled `lbl` in the API graph.
```

Comment:
> I guess we are looking for "value parts of a definition".  

Yes, that is one way to put it.  
E.g. for a def-node representing a property-write `obj.foo = bar`, the rhs is `bar`.  

I'm not sure what you mean in the rest of your comment.

Reply:
One "natural" strategy of finding value parts of definitions is to find definitions and extract their value parts. The extraction may be specific to the definition, but there is a priori no beneficial relationship with other definitions and their extractions.

What you do here is different. Here you have a recursive relation for capturing "value parts of definitions" directly, where the value part of one definition is defined in terms of value parts of other definitions. I was just curious as to what led you to this design.

Reply:
> I was just curious as to what led you to this design. 

The simple answer is that I reused the design from JS.   

The API-graph is recursive, so you have to do something recursive like this in order to construct an API-graph where we can use type-tracking to track a value between the nodes in the API-graph. 

> What you do here is different. Here you have a recursive relation for capturing "value parts of definitions" directly.  

The only thing recursive in this predicate is the construction of the nodes of the API-graph (there are 2 different `rhs()` relations, that might confuse you).

Reply:
I believe that `rhs/3` (defined here) calls `rhs/2` which calls `MkDef` which calls `rhs/3`. If the predicate is in some sense only partially recursive, as you say, perhaps it could indeed be beneficial to refactor it?

Reply:
> perhaps it could indeed be beneficial to refactor it? (not that we need to do that now) 

Maybe, but I doubt it.  
I used to same pattern as JS/Ruby, and I think it's beneficial to keep the same pattern across the languages.   
You can try to refactor it if you wan't, but I don't think you'll get anything better.

Reply:
I definitely think that using the same pattern is the right thing to do right now. I just wondered why (and perhaps even if) that is the best pattern here.